propagator/propagator.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propagator
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// TraceContextHeaderName is the HTTP header field for Google Cloud Trace
+// https://cloud.google.com/trace/docs/setup#force-trace
+const TraceContextHeaderName = "x-cloud-trace-context"
+
+// TraceContextHeaderFormat is the regular expression pattan for valid Cloud Trace header value
+const TraceContextHeaderFormat = "(?P<trace_id>[0-9a-f]{32})/(?P<span_id>[0-9]{1,20})(;o=(?P<trace_flags>[0-9]))?"
+
+// TraceContextHeaderRe is a regular expression object of TraceContextHeaderFormat.
+var TraceContextHeaderRe = regexp.MustCompile(TraceContextHeaderFormat)
+
+var traceContextHeaders = []string{TraceContextHeaderName}
+
+type CloudTraceFormatPropagator struct{}
+
+func New() propagation.TextMapPropagator {
+	return CloudTraceFormatPropagator{}
+}
+
+func (p CloudTraceFormatPropagator) getHeaderValue(carrier propagation.TextMapCarrier) string {
+	header := carrier.Get(TraceContextHeaderName)
+	if header != "" {
+		return header
+	}
+
+	for _, key := range carrier.Keys() {
```

Comment:
It might be worth adding a comment explaining that this is to handle different casings of the header name if not initially found. It's not immediately obvious otherwise

Reply:
Yes, you are right! I added the comment here.